core/icons/mutator_icon.ts

```diff
@@ -0,0 +1,358 @@
+   * the block.
+   */
+  private createMiniWorkspaceChangeListener() {
+    return (e: Abstract) => {
+      if (!this.shouldIgnoreMutatorEvent(e) && !this.updateWorkspacePid) {
+        this.updateWorkspacePid = setTimeout(() => {
+          this.updateWorkspacePid = null;
+          this.recomposeSourceBlock();
+        }, 0);
+      }
+    };
+  }
+
+  /**
+   * Returns true if the given event is not one the mutator needs to
+   * care about.
+   */
+  private shouldIgnoreMutatorEvent(e: Abstract) {
+    return (
+      e.isUiEvent ||
+      e.type === eventUtils.CREATE ||
+      (e.type === eventUtils.CHANGE &&
+        (e as BlockChange).element === 'disabled')
+    );
+  }
+
+  /** Recomposes the source block based on changes to the mini workspace. */
+  private recomposeSourceBlock() {
+    if (!this.rootBlock) return;
+
+    const existingGroup = eventUtils.getGroup();
+    if (!existingGroup) eventUtils.setGroup(true);
+
+    const oldExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+    this.sourceBlock.compose!(this.rootBlock);
+    const newExtraState = BlockChange.getExtraBlockState_(this.sourceBlock);
+
+    if (oldExtraState !== newExtraState) {
+      eventUtils.fire(
+        new (eventUtils.get(eventUtils.BLOCK_CHANGE))(
+          this.sourceBlock,
+          'mutation',
+          null,
+          oldExtraState,
+          newExtraState
+        )
+      );
+    }
+
+    eventUtils.setGroup(existingGroup);
+  }
+
+  /** @internal */
+  getWorkspace(): WorkspaceSvg | undefined {
+    return this.miniWorkspaceBubble?.getWorkspace();
+  }
+
+  /**
+   * Reconnects the given connection to the mutated input on the given block.
+   */
+  static reconnect(
+    connectionChild: Connection,
+    block: Block,
+    inputName: string
+  ): boolean {
+    if (!connectionChild || !connectionChild.getSourceBlock().workspace) {
+      return false; // No connection or block has been deleted.
+    }
+    const connectionParent = block.getInput(inputName)!.connection;
+    const currentParent = connectionChild.targetBlock();
+    if (
+      (!currentParent || currentParent === block) &&
+      connectionParent &&
+      connectionParent.targetConnection !== connectionChild
+    ) {
+      if (connectionParent.isConnected()) {
+        // There's already something connected here.  Get rid of it.
+        connectionParent.disconnect();
+      }
+      connectionParent.connect(connectionChild);
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Returns the parent workspace of a workspace that is inside a mutator,
+   * taking into account wither it is a flyout.
+   */
+  static findParentWs(workspace: WorkspaceSvg): WorkspaceSvg | null {
+    let outerWs = null;
+    if (workspace && workspace.options) {
+      const parent = workspace.options.parentWorkspace;
+      // If we were in a flyout in a mutator, need to go up two levels to find
+      // the actual parent.
```

Comment:
Is this dealing with the case that `workspace` is for the toolbox in a bubble?  Are there other doubly-nested workspaces to worry about?

(And is this function and it's usage correct in the situation of a toolbox-in-a-bubble on a  block that has a `MutatorIcon` _which is itself in a mutator bubble_?)

Reply:
I think this is the only doubly nested workspace case.